pandas/core/groupby/generic.py

```diff
@@ -1213,171 +1213,112 @@ def _wrap_applied_output(self, keys, values, not_indexed_same=False):
         if len(keys) == 0:
             return self.obj._constructor(index=keys)
 
-        key_names = self.grouper.names
-
-        # GH12824
+        # GH12824 - If first value is None, can't assume all are None
         first_not_none = next(com.not_none(*values), None)
 
         if first_not_none is None:
-            # GH9684. If all values are None, then this will throw an error.
-            # We'd prefer it return an empty dataframe.
+            # GH9684 - All values are None, return an empty frame.
             return self.obj._constructor()
-        elif isinstance(first_not_none, DataFrame):
+
+        if isinstance(first_not_none, DataFrame):
             return self._concat_objects(keys, values, not_indexed_same=not_indexed_same)
-        else:
-            if len(self.grouper.groupings) > 1:
-                key_index = self.grouper.result_index
 
+        if isinstance(first_not_none, NDFrame):
+
+            # this is to silence a DeprecationWarning
+            # TODO: Remove when default dtype of empty Series is object
+            kwargs = first_not_none._construct_axes_dict()
+            if isinstance(first_not_none, Series):
+                backup = create_series_with_explicit_dtype(
+                    **kwargs, dtype_if_empty=object
+                )
             else:
-                ping = self.grouper.groupings[0]
-                if len(keys) == ping.ngroups:
-                    key_index = ping.group_index
-                    key_index.name = key_names[0]
+                backup = first_not_none._constructor(**kwargs)
 
-                    key_lookup = Index(keys)
-                    indexer = key_lookup.get_indexer(key_index)
+            values = [x if (x is not None) else backup for x in values]
 
-                    # reorder the values
-                    values = [values[i] for i in indexer]
+        key_index = self.grouper.result_index if self.as_index else None
+        v = values[0]
 
-                    # update due to the potential reorder
-                    first_not_none = next(com.not_none(*values), None)
-                else:
+        if not isinstance(v, (np.ndarray, Index, Series)):
```

Comment:
Can this just be an else statement? Or are there more types we handle than these + NDFrame?

Reply:
I think you're suggesting changing this to read 

    if isinstance(v, (np.ndarray, Index, Series)):
        ...
    else:
        ...

Doing so means the rest of this method is nested within the first `if` clause, which is what I was trying to avoid. Having this nesting separates the else from the if, and reading that deeply nested structure is

Reply:
I think you're suggesting something like:

````
if isinstance(v, (np.ndarray, Index, Series)):
    ...
else:
    ...
````

The reason I have opted not to do this is that the if-block is exceedingly long, whereas the else-block is quite short. Doing it this way would result in a more nested rather than flat structure.

Reply:
Ack - sorry, I see what you're saying now. Ignore my previous response, will investigate.